### DISCUSSION TITLE
Inquiry about TCS Real Car Communication - Handling Point Updates and Resource Release

### DISCUSSION THREAD

**Discussion by BenYaho** [openTCS/opentcs#97] [Q&A] [answered]
HI,

We need guidance on handling point updates and resource release during real car communication in the TCS system. Points are being updated for tasks, but resources are not released, leading to conflicts. Please provide documentation and best practices to resolve this efficiently.
![image](https://github.com/openTCS/opentcs/assets/128113709/a0f03c02-eee1-4d28-bf76-7112795f3de0)

Thank you

**Comment by swltr** [comment]
@BenYaho Resources are released automatically, as long as the communication adapter calls `getProcessModel().commandExecuted()`. Does your communication adapter do that?

**Reply by BenYaho** [reply]
HI,
I think that getProcessModel().commandExecuted() is meant to be used when the entire command is finished. However, what I'm looking for is to release resources after completing a single step. It's like in the diagram where we go from point1 to point4. Currently, my vehicle is at point3. But the resources from point1 to point2 and point2 to point3 haven't been released yet.

To address this issue, I would appreciate guidance on how to resolve it. It seems that the current implementation doesn't release resources for each individual step, and I'm unsure about the appropriate approach to achieve this. 


Thank you.

![image](https://github.com/openTCS/opentcs/assets/128113709/ecf0ae8b-c50a-411d-a399-d43cdc2f65fb)

**Reply by swltr** [reply, marked as answer]
`commandExecuted()` is to be called with the `MovementCommand` that was completed. A movement command is a single step on the vehicle's route. In the situation in your screenshot, there would be the following steps / movement commands:

1. Movement from Point-0001 to Point-0002
2. Movement from Point-0002 to Point-0003
3. Movement from Point-0003 to Point-0004

The vehicle seems to have already completed the first two steps, but the vehicle driver probably only reported the position updates and did not call `commandExecuted()`, so openTCS did not free the resources occupied by the vehicle.
